coreneuron/nrniv/nrn_acc_manager.cpp

```diff
@@ -42,371 +43,465 @@ void setup_nrnthreads_on_device(NrnThread* threads, int nthreads) {
-     */
-    d_threads = (NrnThread*)acc_copyin(threads, sizeof(NrnThread) * nthreads);
-
-    if (interleave_info == NULL) {
-        printf("\n Warning: No permutation data? Required for linear algebra!");
+    if (nrn_ion_global_map_size) {
+        double** d_data =
+            (double**)acc_copyin(nrn_ion_global_map, sizeof(double*) * nrn_ion_global_map_size);
+        for (int j = 0; j < nrn_ion_global_map_size; j++) {
+            if (nrn_ion_global_map[j]) {
+                /* @todo: fix this constant size 3 :( */
+                double* d_mechmap = (double*)acc_copyin(nrn_ion_global_map[j], 3 * sizeof(double));
+                acc_memcpy_to_device(&(d_data[j]), &d_mechmap, sizeof(double*));
+            }
+        }
     }
 
-    /* pointers for data struct on device, starting with d_ */
+    //(NrnThread*)acc_copyin(threads, sizeof(NrnThread) * nthreads);
+    d_threads = threads;
 
     for (i = 0; i < nthreads; i++) {
-        NrnThread* nt = threads + i;      // NrnThread on host
-        NrnThread* d_nt = d_threads + i;  // NrnThread on device
+        NrnThread *nt = threads + i;      // NrnThread on host
+        NrnThread *d_nt = d_threads + i;  // NrnThread on device
 
-        if (nt->end <= 0) {
-            // this is an empty thread and could be only artificial
-            // cells. In this case nothing is executed on gpu.
-            nt->compute_gpu = 0;
-            continue;
+        if (nt->n_presyn) {
+           PreSyn* d_presyns = (PreSyn*)acc_copyin(nt->presyns, sizeof(PreSyn) * nt->n_presyn);
         }
 
-        double* d__data;  // nrn_threads->_data on device
+        if (nt->n_vecplay) {
+        /* copy VecPlayContinuous instances */
+
+            printf("\n Warning: VectorPlay used but NOT implemented on GPU! ");
 
-        /* -- copy _data to device -- */
+            /** just empty containers */
+            void** d_vecplay = (void**)acc_copyin(nt->_vecplay, sizeof(void*) * nt->n_vecplay);
+            // note: we are using unified memory for NrnThread. Once VecPlay is copied to gpu,
+            // we dont want to update nt->vecplay because it will also set gpu pointer of vecplay
+            // inside nt on cpu (due to unified memory).
 
-        /*copy all double data for thread */
-        d__data = (double*)acc_copyin(nt->_data, nt->_ndata * sizeof(double));
+            //acc_memcpy_to_device(&(d_nt->_vecplay), &d_vecplay, sizeof(void**));
 
-        /* Here is the example of using OpenACC data enter/exit
-         * Remember that we are not allowed to use nt->_data but we have to use:
-         *      double *dtmp = nt->_data;  // now use dtmp!
-                #pragma acc enter data copyin(dtmp[0:nt->_ndata]) async(nt->stream_id)
-                #pragma acc wait(nt->stream_id)
-         */
+            for (int i = 0; i < nt->n_vecplay; i++) {
+                VecPlayContinuous* vecplay_instance = (VecPlayContinuous*)nt->_vecplay[i];
```

Comment:
VecPlay code is duplicated in unified and non-unified function. Turn that into function?